<!DOCTYPE html> <!--[if lt IE 7]><html class="no-js lt-ie9 lt-ie8 lt-ie7"> <![endif]--> <!--[if (IE 7)&!(IEMobile)]><html class="no-js lt-ie9 lt-ie8"><![endif]--> <!--[if (IE 8)&!(IEMobile)]><html class="no-js lt-ie9"><![endif]--> <!--[if gt IE 8]><!--> <html class="no-js"><!--<![endif]--> <head> <meta charset="UTF-8"> <meta content="text/html; charset=UTF-8" http-equiv="Content-Type"> <meta http-equiv=X-UA-Compatible content="IE=edge,chrome=1"> <title>A Simple Epigenetic Clock Using Python and SciKit-Learn, BIG Summer Tutorial &#8211; In the Helix</title> <meta name="description" content="In the Helix is a place to post my thoughts on science and other things."> <meta name="keywords" content="Tutorial, Biomarker, Methylation"> <!-- Twitter Cards --> <meta name="twitter:card" content="summary"> <meta name="twitter:image" content="http://localhost:4000/assets/img/new_logo.png"> <meta name="twitter:title" content="A Simple Epigenetic Clock Using Python and SciKit-Learn, BIG Summer Tutorial"> <meta name="twitter:description" content="Brief Tutorial on Fitting a DNA Methylation Based Epigenetic Clock for the Bruins In Genomics (BIG) Summer Program"> <meta name="twitter:site" content="@CpGenome"> <meta name="twitter:creator" content="@CpGenome"> <!-- Open Graph --> <meta property="og:locale" content="en_US"> <meta property="og:type" content="article"> <meta property="og:title" content="A Simple Epigenetic Clock Using Python and SciKit-Learn, BIG Summer Tutorial"> <meta property="og:description" content="Brief Tutorial on Fitting a DNA Methylation Based Epigenetic Clock for the Bruins In Genomics (BIG) Summer Program"> <meta property="og:url" content="http://localhost:4000/EpigeneticClockExample/"> <meta property="og:site_name" content="In the Helix"> <meta property="og:image" content="http://localhost:4000/assets/img/new_logo.png"> <link rel="canonical" href="http://localhost:4000/EpigeneticClockExample/"> <link href="http://localhost:4000/feed.xml" type="application/atom+xml" rel="alternate" title="In the Helix Feed"> <!-- Handheld --> <meta name="HandheldFriendly" content="True"> <meta name="MobileOptimized" content="320"> <meta name="viewport" content="width=device-width, initial-scale=1.0"> <!-- CSS --> <link rel="stylesheet" href="http://localhost:4000/assets/css/main.css"> <!-- JS --> <script src="http://localhost:4000/assets/js/modernizr-3.3.1.custom.min.js"></script> <!-- Favicons --> <link rel="apple-touch-icon" href="http://localhost:4000/assets/img/favicons/apple-icon-precomposed.png"> <link rel="apple-touch-icon" sizes="72x72" href="http://localhost:4000/assets/img/favicons/apple-icon-72x72.png"> <link rel="apple-touch-icon" sizes="114x114" href="http://localhost:4000/assets/img/favicons/apple-icon-114x114.png"> <link rel="apple-touch-icon" sizes="144x144" href="http://localhost:4000/assets/img/favicons/apple-icon-144x144.png"> <link rel="shortcut icon" type="image/png" href="http://localhost:4000/favicon.png" /> <link rel="shortcut icon" href="http://localhost:4000/favicon.ico" /> <!-- Background Image --> <style type="text/css">body {background-image:url(http://localhost:4000/assets/img/background.jpg); background-repeat: no-repeat; background-size: cover; }</style> <!-- Post Feature Image --> </head> <body> <nav id="dl-menu" class="dl-menuwrapper" role="navigation"> <button class="dl-trigger">Open Menu</button> <ul class="dl-menu"> <li><a href="http://localhost:4000/">Home</a></li> <li> <a href="#">About</a> <ul class="dl-submenu"> <li> <img src="http://localhost:4000/assets/img/new_logo.png" alt="In the Helix photo" class="author-photo"> <h4>In the Helix</h4> <p>In the Helix is a place to post my thoughts on science and other things.</p> </li> <li><a href="http://localhost:4000/about/"><span class="btn btn-inverse">Learn More</span></a></li> <li> <a href="http://twitter.com/CpGenome" target="_blank" rel="noopener noreferrer"><i class="fa fa-fw fa-twitter-square"></i> Twitter</a> </li> <li> <a href="http://github.com/NuttyLogic" target="_blank" rel="noopener noreferrer"><i class="fa fa-fw fa-github"></i> Github</a> </li> </ul><!-- /.dl-submenu --> </li> <li> <a href="#">Posts</a> <ul class="dl-submenu"> <li><a href="http://localhost:4000/posts/">All Posts</a></li> <li><a href="http://localhost:4000/tags/">All Tags</a></li> </ul> </li> <li><a href="http://localhost:4000/projects/" >Tools</a></li> <li><a href="http://localhost:4000/teaching/" >BIG Summer</a></li> </ul><!-- /.dl-menu --> </nav><!-- /.dl-menuwrapper --> <!-- Header --> <header class="header" role="banner"> <div class="wrapper animated fadeIn"> <div class="content"> <div class="post-title "> <h1>A Simple Epigenetic Clock Using Python and SciKit-Learn, BIG Summer Tutorial</h1> <h4>19 Jun 2018</h4> <p class="reading-time"> <i class="fa fa-clock-o"></i> Reading time ~7 minutes </p><!-- /.entry-reading-time --> <a class="btn zoombtn" href="http://localhost:4000/teaching/"> <i class="fa fa-chevron-left"></i> </a> </div> <h1 id="epigentic-biomarker-to-predict-age">Epigentic Biomarker to Predict Age</h1> <p>DNA methylation has emerged as a useful proxy for measuring the physiological state of an organism. DNA methylation is dynamic, changing over time in response to environmental stimuli, yet can also be stable for relatively long periods of time<a href="https://www.ncbi.nlm.nih.gov/pubmed/20395474">1</a>. Thus, an individual who continuously lacked exercise and ate poorly for years would have an epigenetic profile that reflected this continued behavior. The dynamic yet stable nature of DNA methylation is extremely beneficial, maintaining information lost with more transitory signals, such as gene expression. This makes DNA methylation ideal for the development of biomarkers, which have already been developed to assess age <a href="https://www.ncbi.nlm.nih.gov/pubmed/24138928">2</a>,<a href="https://www.ncbi.nlm.nih.gov/pubmed/23177740">3</a> and BMI <a href="'Epigenome-wide association study of body mass index, and the adverse outcomes of adiposity.'">4</a>.</p> <p>In this example we will use publicly available data from <a href="https://www.ncbi.nlm.nih.gov/pubmed/23034122">Aging effects on DNA methylation modules in human brain and blood tissue, 5</a> to fit an epigenetic clock using a penalized regression model. The data from this paper can be accessed at <a href="https://www.ncbi.nlm.nih.gov/geo/query/acc.cgi?acc=GSE41169">GSE41169</a>. We will be working with the <a href="ftp://ftp.ncbi.nlm.nih.gov/geo/series/GSE41nnn/GSE41169/matrix/GSE41169_series_matrix.txt.gz"><em>series_matrix_file</em></a> deposited in GEO repository. The complete workflow will utilize several tools:</p> <ol> <li><a href="http://scikit-learn.org/stable/index.html">SciKit-Learn</a>; Several modules from SciKit will be utilized in this example <ol> <li><a href="http://scikit-learn.org/stable/modules/generated/sklearn.decomposition.PCA.html">Principal Component Analysis (PCA)</a>, used during quality control to identify and sample outliers</li> <li><a href="http://scikit-learn.org/stable/modules/generated/sklearn.model_selection.train_test_split.html">Train Test Split</a>, this provides a convenient way to split data into training and testing sets</li> <li><a href="http://scikit-learn.org/stable/modules/generated/sklearn.linear_model.LassoCV.html">Lasso, Cross Validated</a>, this is an implementation of a penalized regression model that we will us to generate an epigentic age model</li> </ol> </li> <li><a href="https://pandas.pydata.org/">Pandas</a>, this library will help us organize data for downstream analysis</li> <li>Numerical Operation Libraries <ol> <li><a href="http://www.numpy.org/">Numpy</a>, used to perform efficient, vectorized mathematical operations and matrix handling</li> <li><a href="https://www.scipy.org/">Scipy</a>, we will use the scipy statistics module to score our regression model</li> </ol> </li> <li>Visualization libraries, to visualize quality control and the model output <ol> <li><a href="https://matplotlib.org/">MatPlotLib</a></li> <li><a href="https://seaborn.pydata.org/">Seaborn</a></li> </ol> </li> </ol> <h2 id="workflow">Workflow</h2> <p>This workflow should be tailored to the data set and phenotype of interest, but the general process should be similar for many biomarkers. I start by setting up an analysis environment in <a href="http://jupyterlab.readthedocs.io/en/stable/">Jupyter</a> and pre-processing the data. After pre-processing the data we will perform quality control followed by fitting the epigentic age model.</p> <h3 id="setting-an-analysis-environment">Setting an Analysis Environment</h3> <p>After a jupyter instance has been created we will setup the analysis environment by import libraries and setting paths.</p> <div class="language-python highlighter-rouge"><div class="highlight"><pre class="highlight"><code><span class="c"># set working directory where the series matrix data is stored</span>
<span class="n">wd</span> <span class="o">=</span> <span class="s">'path to working directory'</span>

<span class="kn">import</span> <span class="nn">pandas</span> <span class="k">as</span> <span class="n">pd</span>
<span class="kn">import</span> <span class="nn">numpy</span> <span class="k">as</span> <span class="n">np</span>
<span class="kn">import</span> <span class="nn">seaborn</span> <span class="k">as</span> <span class="n">sns</span>
<span class="kn">import</span> <span class="nn">matplotlib.pyplot</span> <span class="k">as</span> <span class="n">plt</span>
<span class="kn">from</span> <span class="nn">scipy</span> <span class="kn">import</span> <span class="n">stats</span>
<span class="kn">from</span> <span class="nn">sklearn.decomposition</span> <span class="kn">import</span> <span class="n">PCA</span>
<span class="kn">from</span> <span class="nn">sklearn.linear_model</span> <span class="kn">import</span> <span class="n">LassoCV</span>
<span class="kn">from</span> <span class="nn">sklearn.model_selection</span> <span class="kn">import</span> <span class="n">train_test_split</span>
</code></pre></div></div> <h3 id="pre-processing-data">Pre-Processing Data</h3> <p>The raw data from GEO needs to be processed before a model can be fit. Pre-processing the data will change based on the input and phenotype of interest, however generalizable tools can be written to a handle diverse array of inputs. Once these tools are written, they are often imported into the jupyter environment and not written out in the notebook. However, for this example the tools are written out to highlight data pre-processing.</p> <p>First we define an iterator that takes a text file line, processes that line and returns a list.</p> <div class="language-python highlighter-rouge"><div class="highlight"><pre class="highlight"><code><span class="kn">import</span> <span class="nn">gzip</span>

<span class="k">class</span> <span class="nc">OpenSeriesMatrix</span><span class="p">:</span>
    <span class="s">"""Simple class to iterate over series_matrix_files
    Arguments:
        series (str): path to series file
    Attributes:
        self.f (object): read object
        self.process_line: decodes line if necessary and returns processed list
        self.__iter__: iteration method
        """</span>

    <span class="k">def</span> <span class="nf">__init__</span><span class="p">(</span><span class="bp">self</span><span class="p">,</span> <span class="n">series</span><span class="o">=</span><span class="bp">None</span><span class="p">):</span>
        <span class="c"># if file ends with .gz open as a binary file, else open at txt file</span>
        <span class="k">if</span> <span class="n">series</span><span class="o">.</span><span class="n">endswith</span><span class="p">(</span><span class="s">".gz"</span><span class="p">):</span>
            <span class="bp">self</span><span class="o">.</span><span class="n">f</span> <span class="o">=</span> <span class="n">gzip</span><span class="o">.</span><span class="nb">open</span><span class="p">(</span><span class="n">series</span><span class="p">,</span> <span class="s">'rb'</span><span class="p">)</span>
        <span class="k">else</span><span class="p">:</span>
            <span class="bp">self</span><span class="o">.</span><span class="n">f</span> <span class="o">=</span> <span class="nb">open</span><span class="p">(</span><span class="n">series</span><span class="p">,</span> <span class="s">'r'</span><span class="p">)</span>

    <span class="k">def</span> <span class="nf">__iter__</span><span class="p">(</span><span class="bp">self</span><span class="p">):</span>
        <span class="k">with</span> <span class="bp">self</span><span class="o">.</span><span class="n">f</span> <span class="k">as</span> <span class="n">cg</span><span class="p">:</span>
            <span class="k">while</span> <span class="bp">True</span><span class="p">:</span>
                <span class="n">line</span> <span class="o">=</span> <span class="n">cg</span><span class="o">.</span><span class="n">readline</span><span class="p">()</span>
                <span class="c"># if line is blank break loop</span>
                <span class="k">if</span> <span class="ow">not</span> <span class="n">line</span><span class="p">:</span>
                    <span class="k">break</span>
                <span class="k">yield</span> <span class="bp">self</span><span class="o">.</span><span class="n">process_line</span><span class="p">(</span><span class="n">line</span><span class="p">)</span>
    <span class="err">@</span> <span class="n">Staticmethod</span>
    <span class="k">def</span> <span class="nf">process_line</span><span class="p">(</span><span class="n">line</span><span class="p">):</span>
        <span class="k">if</span> <span class="nb">isinstance</span><span class="p">(</span><span class="n">line</span><span class="p">,</span> <span class="nb">bytes</span><span class="p">):</span>
            <span class="k">return</span> <span class="n">line</span><span class="o">.</span><span class="n">decode</span><span class="p">(</span><span class="s">'utf-8'</span><span class="p">)</span><span class="o">.</span><span class="n">replace</span><span class="p">(</span><span class="s">'</span><span class="se">\n</span><span class="s">'</span><span class="p">,</span> <span class="s">''</span><span class="p">)</span><span class="o">.</span><span class="n">split</span><span class="p">(</span><span class="s">'</span><span class="se">\t</span><span class="s">'</span><span class="p">)</span>
        <span class="k">else</span><span class="p">:</span>
            <span class="k">return</span> <span class="n">line</span><span class="o">.</span><span class="n">replace</span><span class="p">(</span><span class="s">'</span><span class="se">\n</span><span class="s">'</span><span class="p">,</span> <span class="s">''</span><span class="p">)</span><span class="o">.</span><span class="n">split</span><span class="p">(</span><span class="s">'</span><span class="se">\t</span><span class="s">'</span><span class="p">)</span>
</code></pre></div></div> <p>Next we define how to store the data, identifiers present in the series matrix file are passed to the parser which then stores formatted data.</p> <div class="language-python highlighter-rouge"><div class="highlight"><pre class="highlight"><code><span class="k">class</span> <span class="nc">SeriesMatrixParser</span><span class="p">:</span>
    <span class="s">"""Class to parse series matrix files into three components;
    file information, phenotype information, and methylation data.
    Arguments:
        series_matrix_path (str): path to series matrix file
        description_ids (list): list of identifiers for file descriptors
        sample_id (str): identifier for line listing sample names
        phenotype_ids(list): list of identifiers for phenotype information
        matrix_start (str): identifier list the start of the methylation matrix
    Attributes:
        self.series_matrix (OpenSeriesMatrix): iterator object
        self.series_description (dict): dict of file decriptors
        self.sample_ids (list): list of sample names
        self.phenotype_matrix {dict}: dict of phenotype information
        self.matrix_trigger (bool): bool to set beginning of
            methylation matrix in series matrix file
        self.matrix (list): list of methylation sites with values
        self.run (func): wrapper to get series matrix info
    """</span>

    <span class="k">def</span> <span class="nf">__init__</span><span class="p">(</span><span class="bp">self</span><span class="p">,</span> <span class="n">series_matrix_path</span><span class="o">=</span><span class="bp">None</span><span class="p">):</span>
        <span class="k">assert</span><span class="p">(</span><span class="nb">isinstance</span><span class="p">(</span><span class="n">series_matrix_path</span><span class="p">,</span> <span class="nb">str</span><span class="p">))</span>
        <span class="bp">self</span><span class="o">.</span><span class="n">series_matrix</span> <span class="o">=</span> <span class="n">OpenSeriesMatrix</span><span class="p">(</span><span class="n">series_matrix_path</span><span class="p">)</span>
        <span class="bp">self</span><span class="o">.</span><span class="n">series_description</span> <span class="o">=</span> <span class="p">{}</span>
        <span class="bp">self</span><span class="o">.</span><span class="n">sample_ids</span> <span class="o">=</span> <span class="p">[]</span>
        <span class="bp">self</span><span class="o">.</span><span class="n">phenotype_matrix</span> <span class="o">=</span> <span class="p">{}</span>
        <span class="bp">self</span><span class="o">.</span><span class="n">matrix_trigger</span> <span class="o">=</span> <span class="bp">False</span>
        <span class="bp">self</span><span class="o">.</span><span class="n">matrix</span> <span class="o">=</span> <span class="p">[]</span>

    <span class="k">def</span> <span class="nf">run</span><span class="p">(</span><span class="bp">self</span><span class="p">,</span> <span class="n">description_ids</span><span class="o">=</span><span class="bp">None</span><span class="p">,</span> <span class="n">sample_id</span><span class="o">=</span><span class="bp">None</span><span class="p">,</span> <span class="n">phenotype_ids</span><span class="o">=</span><span class="bp">None</span><span class="p">,</span> <span class="n">matrix_start</span><span class="o">=</span><span class="bp">None</span><span class="p">):</span>
        <span class="k">assert</span><span class="p">(</span><span class="nb">isinstance</span><span class="p">(</span><span class="n">description_ids</span><span class="p">,</span> <span class="nb">list</span><span class="p">))</span>
        <span class="k">assert</span><span class="p">(</span><span class="nb">isinstance</span><span class="p">(</span><span class="n">sample_id</span><span class="p">,</span> <span class="nb">str</span><span class="p">))</span>
        <span class="k">assert</span><span class="p">(</span><span class="nb">isinstance</span><span class="p">(</span><span class="n">phenotype_ids</span><span class="p">,</span> <span class="nb">list</span><span class="p">))</span>
        <span class="k">assert</span><span class="p">(</span><span class="nb">isinstance</span><span class="p">(</span><span class="n">matrix_start</span><span class="p">,</span> <span class="nb">str</span><span class="p">))</span>
        <span class="k">for</span> <span class="n">line</span> <span class="ow">in</span> <span class="bp">self</span><span class="o">.</span><span class="n">series_matrix</span><span class="p">:</span>
            <span class="k">if</span> <span class="ow">not</span> <span class="bp">self</span><span class="o">.</span><span class="n">matrix_trigger</span><span class="p">:</span>
                <span class="bp">self</span><span class="o">.</span><span class="n">get_descriptive_lines</span><span class="p">(</span><span class="n">line</span><span class="p">,</span> <span class="n">description_ids</span><span class="p">)</span>
                <span class="bp">self</span><span class="o">.</span><span class="n">get_sample_ids</span><span class="p">(</span><span class="n">line</span><span class="p">,</span> <span class="n">sample_id</span><span class="p">)</span>
                <span class="bp">self</span><span class="o">.</span><span class="n">get_phenotype_info</span><span class="p">(</span><span class="n">line</span><span class="p">,</span> <span class="n">phenotype_ids</span><span class="p">)</span>
                <span class="bp">self</span><span class="o">.</span><span class="n">get_matrix</span><span class="p">(</span><span class="n">line</span><span class="p">,</span> <span class="n">matrix_start</span><span class="o">=</span><span class="n">matrix_start</span><span class="p">)</span>
            <span class="k">else</span><span class="p">:</span>
                <span class="bp">self</span><span class="o">.</span><span class="n">get_matrix</span><span class="p">(</span><span class="n">line</span><span class="p">)</span>

    <span class="k">def</span> <span class="nf">get_descriptive_lines</span><span class="p">(</span><span class="bp">self</span><span class="p">,</span> <span class="n">line</span><span class="p">,</span> <span class="n">description_ids</span><span class="p">):</span>
        <span class="k">if</span> <span class="n">line</span><span class="p">[</span><span class="mi">0</span><span class="p">]</span> <span class="ow">in</span> <span class="n">description_ids</span><span class="p">:</span>
            <span class="k">try</span><span class="p">:</span>
                <span class="n">info</span> <span class="o">=</span> <span class="bp">self</span><span class="o">.</span><span class="n">series_description</span><span class="p">[</span><span class="n">line</span><span class="p">[</span><span class="mi">0</span><span class="p">]]</span>
            <span class="k">except</span> <span class="nb">KeyError</span><span class="p">:</span>
                <span class="bp">self</span><span class="o">.</span><span class="n">series_description</span><span class="p">[</span><span class="n">line</span><span class="p">[</span><span class="mi">0</span><span class="p">]]</span> <span class="o">=</span> <span class="n">line</span><span class="p">[</span><span class="mi">1</span><span class="p">:]</span>
            <span class="k">else</span><span class="p">:</span>
                <span class="bp">self</span><span class="o">.</span><span class="n">series_description</span><span class="p">[</span><span class="n">line</span><span class="p">[</span><span class="mi">0</span><span class="p">]]</span> <span class="o">=</span> <span class="n">info</span> <span class="o">+</span> <span class="n">line</span><span class="p">[</span><span class="mi">1</span><span class="p">:]</span>

    <span class="k">def</span> <span class="nf">get_sample_ids</span><span class="p">(</span><span class="bp">self</span><span class="p">,</span> <span class="n">line</span><span class="p">,</span> <span class="n">sample_id</span><span class="p">):</span>
        <span class="k">if</span> <span class="n">line</span><span class="p">[</span><span class="mi">0</span><span class="p">]</span> <span class="o">==</span> <span class="n">sample_id</span><span class="p">:</span>
            <span class="bp">self</span><span class="o">.</span><span class="n">sample_ids</span> <span class="o">=</span> <span class="n">line</span><span class="p">[</span><span class="mi">1</span><span class="p">:]</span>

    <span class="k">def</span> <span class="nf">get_phenotype_info</span><span class="p">(</span><span class="bp">self</span><span class="p">,</span> <span class="n">line</span><span class="p">,</span> <span class="n">phenotype_ids</span><span class="p">):</span>
        <span class="k">if</span> <span class="n">line</span><span class="p">[</span><span class="mi">0</span><span class="p">]</span> <span class="ow">in</span> <span class="n">phenotype_ids</span><span class="p">:</span>
            <span class="n">phenotype_label</span> <span class="o">=</span> <span class="n">line</span><span class="p">[</span><span class="mi">1</span><span class="p">]</span><span class="o">.</span><span class="n">split</span><span class="p">(</span><span class="s">':'</span><span class="p">)[</span><span class="mi">0</span><span class="p">]</span><span class="o">.</span><span class="n">strip</span><span class="p">(</span><span class="s">' "'</span><span class="p">)</span>
            <span class="bp">self</span><span class="o">.</span><span class="n">phenotype_matrix</span><span class="p">[</span><span class="n">phenotype_label</span><span class="p">]</span> <span class="o">=</span> <span class="p">[]</span>
            <span class="k">for</span> <span class="n">phenotype</span> <span class="ow">in</span> <span class="n">line</span><span class="p">[</span><span class="mi">1</span><span class="p">:]:</span>
                <span class="n">phenotype_split</span> <span class="o">=</span> <span class="n">phenotype</span><span class="o">.</span><span class="n">split</span><span class="p">(</span><span class="s">':'</span><span class="p">)</span>
                <span class="bp">self</span><span class="o">.</span><span class="n">phenotype_matrix</span><span class="p">[</span><span class="n">phenotype_label</span><span class="p">]</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="n">phenotype_split</span><span class="p">[</span><span class="mi">1</span><span class="p">]</span><span class="o">.</span><span class="n">strip</span><span class="p">(</span><span class="s">' "'</span><span class="p">))</span>

    <span class="k">def</span> <span class="nf">get_matrix</span><span class="p">(</span><span class="bp">self</span><span class="p">,</span> <span class="n">line</span><span class="p">,</span> <span class="n">matrix_start</span><span class="o">=</span><span class="bp">None</span><span class="p">):</span>
        <span class="k">if</span> <span class="bp">self</span><span class="o">.</span><span class="n">matrix_trigger</span><span class="p">:</span>
            <span class="bp">self</span><span class="o">.</span><span class="n">matrix</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="n">line</span><span class="p">)</span>
        <span class="k">elif</span> <span class="n">line</span><span class="p">[</span><span class="mi">0</span><span class="p">]</span> <span class="o">==</span> <span class="n">matrix_start</span><span class="p">:</span>
            <span class="bp">self</span><span class="o">.</span><span class="n">matrix_trigger</span> <span class="o">=</span> <span class="bp">True</span>
</code></pre></div></div> <p>We now want to initialize the classes above with our series matrix data, specifying the file lines we are interested in using line identifiers. The line identifiers are selected by manually inspecting the file. Additionally, we will transform the returned methylation values into a pandas dataframe and perform operations on the dataframe to simplify downstream analysis.</p> <div class="language-python highlighter-rouge"><div class="highlight"><pre class="highlight"><code><span class="c"># name geo file</span>
<span class="n">geo_file</span> <span class="o">=</span> <span class="s">'GSE41169_series_matrix.txt'</span>

<span class="c"># run parser class on the downloaded information, you will have to</span>
<span class="c"># identify phenotype information and descriptors manually</span>
<span class="n">example_matrix</span> <span class="o">=</span> <span class="n">SeriesMatrixParser</span><span class="p">(</span><span class="n">f</span><span class="s">'{wd}{geo_file}'</span><span class="p">)</span>
<span class="n">example_matrix</span><span class="o">.</span><span class="n">run</span><span class="p">(</span><span class="n">description_ids</span><span class="o">=</span><span class="p">[</span><span class="s">'!Series_title'</span><span class="p">,</span> <span class="s">'!Series_geo_accession'</span><span class="p">,</span>
                                    <span class="s">'!Series_pubmed_id'</span><span class="p">,</span> <span class="s">'!Series_summary'</span><span class="p">,</span>
                                    <span class="s">'!Series_overall_design'</span><span class="p">,</span>
                                    <span class="s">'!Series_sample_id'</span><span class="p">,</span> <span class="s">'!Series_relation'</span><span class="p">],</span>
               <span class="n">sample_id</span><span class="o">=</span><span class="s">'!Sample_geo_accession'</span><span class="p">,</span>
               <span class="n">phenotype_ids</span><span class="o">=</span><span class="p">[</span><span class="s">'!Sample_characteristics_ch1'</span><span class="p">],</span>
               <span class="n">matrix_start</span><span class="o">=</span><span class="s">'!series_matrix_table_begin'</span><span class="p">)</span>

<span class="c"># transform matrix list into a pandas dataframe</span>
<span class="n">example_matrix_df</span> <span class="o">=</span> <span class="n">pd</span><span class="o">.</span><span class="n">DataFrame</span><span class="p">(</span><span class="n">data</span><span class="o">=</span><span class="n">example_matrix</span><span class="o">.</span><span class="n">matrix</span><span class="p">[</span><span class="mi">1</span><span class="p">:</span><span class="o">-</span><span class="mi">1</span><span class="p">],</span>
                                <span class="n">columns</span><span class="o">=</span><span class="n">example_matrix</span><span class="o">.</span><span class="n">matrix</span><span class="p">[</span><span class="mi">0</span><span class="p">])</span>

<span class="c"># set index</span>
<span class="n">example_matrix_df</span> <span class="o">=</span> <span class="n">example_matrix_df</span><span class="o">.</span><span class="n">set_index</span><span class="p">(</span><span class="s">'"ID_REF"'</span><span class="p">)</span>

<span class="c"># transform strings to float values</span>
<span class="n">example_matrix_df</span> <span class="o">=</span> <span class="n">example_matrix_df</span><span class="o">.</span><span class="nb">apply</span><span class="p">(</span><span class="n">pd</span><span class="o">.</span><span class="n">to_numeric</span><span class="p">,</span> <span class="n">errors</span><span class="o">=</span><span class="s">'coerce'</span><span class="p">)</span>

<span class="c"># drop rows, methylation sites, with missing infromation</span>
<span class="n">example_matrix_df</span> <span class="o">=</span> <span class="n">example_matrix_df</span><span class="o">.</span><span class="n">dropna</span><span class="p">(</span><span class="n">axis</span><span class="o">=</span><span class="mi">0</span><span class="p">)</span>
</code></pre></div></div> <p>The final step of pre-processing is retrieving the phenotype of interest, in this case age.</p> <div class="language-python highlighter-rouge"><div class="highlight"><pre class="highlight"><code><span class="c"># retrieve age phenotype</span>
<span class="n">example_matrix_age</span> <span class="o">=</span> <span class="p">[</span><span class="nb">int</span><span class="p">(</span><span class="n">x</span><span class="p">)</span> <span class="k">for</span> <span class="n">x</span> <span class="ow">in</span> <span class="n">example_matrix</span><span class="o">.</span><span class="n">phenotype_matrix</span><span class="p">[</span><span class="s">'age'</span><span class="p">]]</span>
</code></pre></div></div> <h2 id="matrix-quality-control">Matrix Quality Control</h2> <p>To ensure there aren’t large technical biases between samples, we will decompose the matrix using PCA then graph the output to check for outliers. If there are outliers we will remove them from downstream analysis.</p> <div class="language-python highlighter-rouge"><div class="highlight"><pre class="highlight"><code><span class="c"># define a PCA object</span>
<span class="n">qc_pca</span> <span class="o">=</span> <span class="n">PCA</span><span class="p">(</span><span class="n">n_components</span><span class="o">=</span><span class="mi">4</span><span class="p">,</span> <span class="n">whiten</span><span class="o">=</span><span class="bp">False</span><span class="p">)</span>

<span class="c"># fit the PCA object</span>
<span class="n">qc_pca_values</span> <span class="o">=</span> <span class="n">qc_pca</span><span class="o">.</span><span class="n">fit_transform</span><span class="p">(</span><span class="n">example_matrix_df</span><span class="o">.</span><span class="n">values</span><span class="o">.</span><span class="n">T</span><span class="p">)</span>

<span class="c"># get the variance explained for the first two principal components</span>
<span class="n">variance_explained</span> <span class="o">=</span> <span class="n">qc_pca</span><span class="o">.</span><span class="n">explained_variance_ratio_</span>

<span class="n">pc1</span> <span class="o">=</span> <span class="n">qc_pca_values</span><span class="p">[:,</span><span class="mi">0</span><span class="p">]</span>
<span class="n">pc2</span> <span class="o">=</span> <span class="n">qc_pca_values</span><span class="p">[:,</span><span class="mi">1</span><span class="p">]</span>
</code></pre></div></div> <p>After decomposing the matrix we graph the matrix, and remove outliers based on a previous run.</p> <div class="language-python highlighter-rouge"><div class="highlight"><pre class="highlight"><code><span class="c"># scatter plot of first two PCs, and retrieve and sample outliers</span>
<span class="c"># list to store sample outliers</span>
<span class="n">non_outlier_list</span> <span class="o">=</span> <span class="p">[]</span>
<span class="n">non_outlier_age</span> <span class="o">=</span> <span class="p">[]</span>
<span class="n">fig</span><span class="p">,</span> <span class="n">ax</span> <span class="o">=</span> <span class="n">plt</span><span class="o">.</span><span class="n">subplots</span><span class="p">(</span><span class="n">figsize</span><span class="o">=</span><span class="p">(</span><span class="mi">12</span><span class="p">,</span><span class="mi">12</span><span class="p">))</span>
<span class="n">ax</span><span class="o">.</span><span class="n">scatter</span><span class="p">(</span><span class="n">pc1</span><span class="p">,</span> <span class="n">pc2</span><span class="p">,)</span>
<span class="n">ax</span><span class="o">.</span><span class="n">set_title</span><span class="p">(</span><span class="s">'Methylation Matrix PCA'</span><span class="p">)</span>
<span class="n">ax</span><span class="o">.</span><span class="n">set_xlabel</span><span class="p">(</span><span class="n">f</span><span class="s">'PC1 Variance Explained = {variance_explained[0]:0.3f}'</span><span class="p">)</span>
<span class="n">ax</span><span class="o">.</span><span class="n">set_ylabel</span><span class="p">(</span><span class="n">f</span><span class="s">'PC2 Variance Explained = {variance_explained[1]:0.3f}'</span><span class="p">)</span>
<span class="c"># iterate through pc1, pc2, and sample labels to add labels to plotted points</span>
<span class="k">for</span> <span class="n">x</span><span class="p">,</span> <span class="n">y</span><span class="p">,</span> <span class="n">label</span><span class="p">,</span> <span class="n">age</span> <span class="ow">in</span> <span class="nb">zip</span><span class="p">(</span><span class="n">pc1</span><span class="p">,</span> <span class="n">pc2</span><span class="p">,</span> <span class="nb">list</span><span class="p">(</span><span class="n">example_matrix_df</span><span class="p">),</span> <span class="n">example_matrix_age</span><span class="p">):</span>
    <span class="n">ax</span><span class="o">.</span><span class="n">text</span><span class="p">(</span><span class="n">x</span><span class="o">=</span><span class="n">x</span><span class="p">,</span> <span class="n">y</span><span class="o">=</span><span class="n">y</span><span class="p">,</span> <span class="n">s</span><span class="o">=</span><span class="n">label</span><span class="p">)</span>
    <span class="c"># add outliers to list</span>
    <span class="k">if</span> <span class="n">x</span> <span class="o">&lt;</span> <span class="mi">5</span><span class="p">:</span>
        <span class="n">non_outlier_list</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="n">label</span><span class="p">)</span>
        <span class="n">non_outlier_age</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="n">age</span><span class="p">)</span>
<span class="n">plt</span><span class="o">.</span><span class="n">show</span><span class="p">()</span>
</code></pre></div></div> <p><img src="https://github.com/NuttyLogic/NuttyLogic.github.io/blob/master/posts/post_assets/epigenetic_clock_example/output_25_0.png?raw=true" alt="" /></p> <h2 id="fit-penalized-regression-model">Fit Penalized Regression Model</h2> <p>With the processed data in hand, a penalized linear regression model can be fit. First we want to dived our data into training and testing sets.</p> <div class="language-python highlighter-rouge"><div class="highlight"><pre class="highlight"><code><span class="c"># want a list of sample names</span>
<span class="n">sample_ids</span> <span class="o">=</span> <span class="nb">list</span><span class="p">(</span><span class="n">example_matrix_df</span><span class="p">[</span><span class="n">non_outlier_list</span><span class="p">])</span>

<span class="n">X_train</span><span class="p">,</span> <span class="n">X_test</span><span class="p">,</span> <span class="n">y_train</span><span class="p">,</span> <span class="n">y_test</span> <span class="o">=</span> <span class="n">train_test_split</span><span class="p">(</span><span class="n">sample_ids</span><span class="p">,</span> <span class="n">non_outlier_age</span><span class="p">,</span> <span class="n">test_size</span><span class="o">=</span><span class="mf">0.1</span><span class="p">)</span>

<span class="c"># take dataframe values as a numpy array and transpose the array with .T</span>
<span class="n">X_train</span> <span class="o">=</span> <span class="n">example_matrix_df</span><span class="p">[</span><span class="n">X_train</span><span class="p">]</span><span class="o">.</span><span class="n">values</span><span class="o">.</span><span class="n">T</span>
<span class="n">X_test</span> <span class="o">=</span> <span class="n">example_matrix_df</span><span class="p">[</span><span class="n">X_test</span><span class="p">]</span><span class="o">.</span><span class="n">values</span><span class="o">.</span><span class="n">T</span>
</code></pre></div></div> <p>We then initialize and fit a penalized regression object, LassoCV, to the training data.</p> <div class="language-python highlighter-rouge"><div class="highlight"><pre class="highlight"><code><span class="c"># initialize a penalized regression object</span>
<span class="n">lasso_cv</span> <span class="o">=</span> <span class="n">LassoCV</span><span class="p">(</span><span class="n">cv</span><span class="o">=</span><span class="mi">3</span><span class="p">,</span> <span class="n">n_jobs</span><span class="o">=</span><span class="mi">2</span><span class="p">)</span>

<span class="c"># fit object</span>
<span class="n">lasso_cv</span><span class="o">.</span><span class="n">fit</span><span class="p">(</span><span class="n">X_train</span><span class="p">,</span> <span class="n">y_train</span><span class="p">)</span>
</code></pre></div></div> <h2 id="score-the-model">Score the Model</h2> <p>Using the testing data, we then predict the age using the methylation matrix and compare the predicted age to the actual age and score the model using Pearson’s R^2.</p> <div class="language-python highlighter-rouge"><div class="highlight"><pre class="highlight"><code><span class="k">def</span> <span class="nf">r2</span><span class="p">(</span><span class="n">x</span><span class="p">,</span> <span class="n">y</span><span class="p">):</span>
    <span class="k">return</span> <span class="n">stats</span><span class="o">.</span><span class="n">pearsonr</span><span class="p">(</span><span class="n">x</span><span class="p">,</span> <span class="n">y</span><span class="p">)[</span><span class="mi">0</span><span class="p">]</span> <span class="o">**</span> <span class="mi">2</span>

<span class="n">predicted_test_age</span> <span class="o">=</span> <span class="n">elastic_net_cv</span><span class="o">.</span><span class="n">predict</span><span class="p">(</span><span class="n">X_test</span><span class="p">)</span>
<span class="n">test_score</span> <span class="o">=</span> <span class="n">r2</span><span class="p">(</span><span class="n">predicted_test_age</span><span class="p">,</span> <span class="n">y_test</span><span class="p">)</span>
</code></pre></div></div> <p>Finally, we plot the model.</p> <div class="language-python highlighter-rouge"><div class="highlight"><pre class="highlight"><code><span class="n">fig</span><span class="p">,</span> <span class="n">ax</span> <span class="o">=</span> <span class="n">plt</span><span class="o">.</span><span class="n">subplots</span><span class="p">(</span><span class="n">figsize</span><span class="o">=</span><span class="p">(</span><span class="mi">12</span><span class="p">,</span><span class="mi">12</span><span class="p">))</span>

<span class="n">ax</span><span class="o">.</span><span class="n">scatter</span><span class="p">(</span><span class="n">predicted_test_age</span><span class="p">,</span> <span class="n">y_test</span><span class="p">,</span> <span class="n">c</span><span class="o">=</span><span class="n">sns</span><span class="o">.</span><span class="n">color_palette</span><span class="p">(</span><span class="s">"Paired"</span><span class="p">)[</span><span class="mi">1</span><span class="p">])</span>
<span class="n">ax</span><span class="o">.</span><span class="n">plot</span><span class="p">([</span><span class="n">np</span><span class="o">.</span><span class="n">asarray</span><span class="p">(</span><span class="n">y_test</span><span class="p">)</span><span class="o">.</span><span class="nb">min</span><span class="p">(),</span> <span class="n">np</span><span class="o">.</span><span class="n">asarray</span><span class="p">(</span><span class="n">y_test</span><span class="p">)</span><span class="o">.</span><span class="nb">max</span><span class="p">()],</span>
<span class="p">[</span><span class="n">np</span><span class="o">.</span><span class="n">asarray</span><span class="p">(</span><span class="n">y_test</span><span class="p">)</span><span class="o">.</span><span class="nb">min</span><span class="p">(),</span> <span class="n">np</span><span class="o">.</span><span class="n">asarray</span><span class="p">(</span><span class="n">y_test</span><span class="p">)</span><span class="o">.</span><span class="nb">max</span><span class="p">()],</span> <span class="s">'k--'</span><span class="p">,</span> <span class="n">lw</span><span class="o">=</span><span class="mi">2</span><span class="p">)</span>
<span class="n">ax</span><span class="o">.</span><span class="n">set_xlabel</span><span class="p">(</span><span class="s">'Predicted Age'</span><span class="p">)</span>
<span class="n">ax</span><span class="o">.</span><span class="n">set_ylabel</span><span class="p">(</span><span class="s">'Actual Age'</span><span class="p">)</span>
<span class="n">ax</span><span class="o">.</span><span class="n">set_title</span><span class="p">(</span><span class="s">'Age Prediction Example'</span><span class="p">)</span>
<span class="n">ax</span><span class="o">.</span><span class="n">text</span><span class="p">(</span><span class="mf">0.01</span><span class="p">,</span> <span class="o">.</span><span class="mi">98</span><span class="p">,</span> <span class="n">f</span><span class="s">'R^2 = {test_score:0.2f}'</span><span class="p">,</span>  <span class="n">transform</span><span class="o">=</span><span class="n">ax</span><span class="o">.</span><span class="n">transAxes</span><span class="p">)</span>

<span class="n">plt</span><span class="o">.</span><span class="n">show</span><span class="p">()</span>
</code></pre></div></div> <p><img src="https://github.com/NuttyLogic/NuttyLogic.github.io/blob/master/posts/post_assets/epigenetic_clock_example/output_41_0.png?raw=true" alt="" /></p> <h3 id="references">References</h3> <ol> <li>Zemach, A., Mcdaniel, I., Silva, P. &amp; Zilberman, D. Genome-Wide Evolutionary Analysis of Eukaryotic DNA Methylation. Science (New York, NY) 11928, science.1186366v1 (2010).</li> <li>Horvath, S. Erratum to: DNA methylation age of human tissues and cell types. Genome Biology 16, 96 (2015).</li> <li>Hannum, G. et al. Genome-wide Methylation Profiles Reveal Quantitative Views of Human Aging Rates. Molecular Cell 49, 359–367 (2013).</li> <li>Wahl, S. et al. Epigenome-wide association study of body mass index, and the adverse outcomes of adiposity. Nature 541, 81–86 (2017).</li> <li>Horvath, S. et al. Aging effects on DNA methylation modules in human brain and blood tissue. Genome Biol. 13, 1465–6914 (2012).</li> </ol> <div class="entry-meta"> <br> <hr> <span class="entry-tags"><a href="http://localhost:4000/tags/#Tutorial" title="Pages tagged Tutorial" class="tag"><span class="term">Tutorial</span></a><a href="http://localhost:4000/tags/#Biomarker" title="Pages tagged Biomarker" class="tag"><span class="term">Biomarker</span></a><a href="http://localhost:4000/tags/#Methylation" title="Pages tagged Methylation" class="tag"><span class="term">Methylation</span></a></span> <span class="social-share"> <a href="https://www.facebook.com/sharer/sharer.php?u=http://localhost:4000/EpigeneticClockExample/" title="Share on Facebook" class="tag"> <span class="term"><i class="fa fa-facebook-square"></i> Share</span> </a> <a href="https://twitter.com/intent/tweet?text=http://localhost:4000/EpigeneticClockExample/" title="Share on Twitter" class="tag"> <span class="term"><i class="fa fa-twitter-square"></i> Tweet</span> </a> <a href="https://plus.google.com/share?url=http://localhost:4000/EpigeneticClockExample/" title="Share on Google+" class="tag"> <span class="term"><i class="fa fa-google-plus-square"></i> +1</span> </a> </span> <div style="clear:both"></div> </div> </div> </div> </header> <!-- JS --> <script src="http://localhost:4000/assets/js/jquery-1.12.0.min.js"></script> <script src="http://localhost:4000/assets/js/jquery.dlmenu.min.js"></script> <script src="http://localhost:4000/assets/js/jquery.goup.min.js"></script> <script src="http://localhost:4000/assets/js/jquery.magnific-popup.min.js"></script> <script src="http://localhost:4000/assets/js/jquery.fitvid.min.js"></script> <script src="http://localhost:4000/assets/js/scripts.js"></script> <!-- MathJax --> <script async src="https://cdn.mathjax.org/mathjax/latest/MathJax.js?config=TeX-AMS-MML_HTMLorMML"></script> </body> </html>
