---
layout: page
type: page
title: Curriculum Vitae
slug: cv
sidebar: true
order: 4
description: 
buttons:
  print: False
  pdf: /assets/cv.pdf
---

## Research Interests

---

- Understanding gene regulation dynamics through the study of epigenetic modifications, mainly DNA methylation.
- Understanding contributors to cancer relapse following immunotherapy and how alternative immunotherapies can slow or prevent relapse.

## Education

---
**University of California, Los Angeles**, Los Angeles, California

Ph.D., Human Genetics, **August 2016 - Spring 2021**

- Advisor: Prof. Matteo Pellegrini
- Thesis Topic: *Assessing human health using DNA methylation based predictive models*
- Thesis Committee: Matteo Pellegrini, Xia Yang, Jae-Hoon Sul, Aldons J. Lusis
- Supported by NIH T32 CA201160-01 Biomedical Big Data (2018 - 2019)

**University of Utah**, Salt Lake City, Utah

M.P.H., Public Health, **August  2011 - June 2013**

- Advisors: Professors Lisa Gren and David Sundwall
- Practicum: *Disconcerting headache practice patterns in academic primary care.*

**University of Utah**, Salt Lake City, Utah

B.S., Chemistry, **August  2006 - December 2010**

## Experience

---
**University of California, Los Angeles**
Graduate Student Researcher, Pellegrini Lab, August 2016 – Current

- Designed an integrated measure of type two diabetes and fit a DNA methylation based predictive model to this measure
- Built a high performance bisulfite sequencing processing platform written in C, C++, and Python
- Developed a novel statistical model of epigenetic aging to model non-linear epigenetic aging and open sourced the model as a Python package  
- Built an analysis pipeline to assess inactivation of CAR T-cell therapy vectors following product infusion using DNA methylation and vector integration location
- Designed a targeted bisulfite sequencing method and implemented an optimized analysis pipeline
- Collaborated with researches, clinicians, and support staff to implement research projects and meet project goals
- Managed several parallel research projects, completing simultaneous tasks and delegating appropriate tasks to assistants
- Carried out large scale data processing on high performance computing clusters and AWS
- Instituted lab wide version control best practice to ease collaboration and foster reproducibility

**University of Utah School of Medicine**
Lab Specialist, Phillips Lab, February 2014 – Current**

- Analyzed sequencing data to identify causative genetic variants in clinical cases diagnosed with iron homeostasis and heme synthesis disorders
- Integrated PacBio, Oxford Nanopore, and Illumina sequencing data to assemble a high quality reference genome for the Babraham pig breed with intact T-Cell Receptor loci for vaccine development
- Performed CRISPR KO analysis to identify ligands targeted by a ubiquitous cancer killing T-Cell line

### Open Source Software Tools

---
**BSBolt** (Python, C, C++)  – High performance bisulfite sequencing processing platform; bsbolt.readthedocs.io
**Epigenetic Pacemaker** (Python) – Evolutionary model of epigenetic aging; epigeneticpacemaker.readthedocs.io

## Computer Skills

---

- Languages
  - Proficient - Python
  - Competent - C, C++, R, Bash
- Unit Testing
- System - Linux/UNIX, MacOS
- Tools - Git, Jupyter, SGE, AWS, SLURM, Anaconda, Vim

## Teaching

---
**Research Immersion Laboratory in Genomic Biology at UCLA**(Spring 2018, Spring 2019) Instructed laboratory class for upper division molecular biology students at UCLA. In the laboratory class I introduced bioinformatics tools and workflow utilized to perform a quarter long genome annotation project. (Review available upon request)

## Mentoring

---

- **Huling Huang (Bruins In Genomics Undergraduate Researcher, Summer 2018):** Modeled non-linear epigenetic aging trends in publicly available datasets and compared epigenetic aging trends across distinct populations.
- **Maria Malikwa (Santa Monica College Summer Scholars, Summer 2018):** Designed a DNA methylation based predictor of folic acid levels in pregnant women.
- **Ibrahime Wone (Santa Monica College Summer Scholars, Summer 2019):** Built a non-linear DNA methylation based predictive model for type two diabetes progression.
- **Adewale Oytetunde (UCLA-CSUN Stem Cell Bridges Program, August 2018 - August 2019):** Modeled methylation dynamics during cellular differentiation and how the DNA methylation landscape is altered by induced pluripotency.
- **Kalsuda Lapborisuth (UCLA Undergraduate Research Fellow, August 2018 - Present)** Developing a simulation framework for modeling epigenetic age acceleration in response to physiological stress and is validating the framework using publicly available methylation datasets.

## Publications

---

- [**Farrell, C.**, Thompson, M., Tosevska, A., Oyetunde, A. & Pellegrini, M. *BiSulfite Bolt: A BiSulfite Sequencing Analysis Platform.* GigaScience  (2021, accepted)](https://www.biorxiv.org/content/10.1101/2020.10.06.328559v2)
- [Li, S., Ma, F., Yokota, T., Jr., G. G., Palermo, A., Wang, Y., **Farrell, C.**, Wang, Y.-C., Wu, R., Zhou, Z., Pan, C., Morselli, M., Teitell, M. A., Ryazantsev, S., Fishbein, G. A., ten Hoeve, J., Arboleda, V. A., Bloom, J., Dillon, B., Pellegrini, M., Lusis, A. J., Graeber, T. G., Arumugaswami, V. & Deb, A. *Metabolic reprogramming and epigenetic changes of vital organs in SARS-CoV-2–induced systemic toxicity.* JCI Insight 6, (2021).](https://insight.jci.org/articles/view/145027)
- [**Farrell, C.**, Snir, S. & Pellegrini, M. *The Epigenetic Pacemaker: modeling epigenetic states under an evolutionary framework.* Bioinformatics 36, 4662–4663 (2020).](https://academic.oup.com/bioinformatics/article-abstract/doi/10.1093/bioinformatics/btaa585/5861533?redirectedFrom=fulltext)
- [Nowicki, T. S., **Farrell, C.**, Morselli, M., Rubbi, L., Campbell, K. M., Macabali, M. H., Berent-Maoz, B., Comin-Anduix, B., Pellegrini, M. & Ribas, A. Epigenetic Suppression of Transgenic T-cell Receptor Expression via Gamma-Retroviral Vector Methylation in Adoptive Cell Transfer Therapy. Cancer Discov. (2020). doi:10.1158/2159-8290.CD-20-0300](https://cancerdiscovery.aacrjournals.org/content/10/11/1645)
- [Morselli, M., **Farrell, C.**, Rubbi, L., Fehling, H. L., Henkhaus, R. & Pellegrini, M. *Targeted bisulfite sequencing for biomarker discovery.* Methods (2020). doi:10.1016/j.ymeth.2020.07.006](https://www.sciencedirect.com/science/article/abs/pii/S1046202320300839)
- [Wu, R., Ma, F., Tosevska, A., **Farrell, C.**, Pellegrini, M. & Deb, A. *Cardiac fibroblast proliferation rates and collagen expression mature early and are unaltered with advancing age.* JCI Insight 5, (2020).](https://df6sxcketz7bb.cloudfront.net/manuscripts/140000/140628/jci.insight.140628.v2.pdf)
- [Levine, M., McDevitt, R. A., Meer, M., Perdue, K., Di Francesco, A., Meade, T., **Farrell, C.**, Thrush, K., Wang, M., Dunn, C., Pellegrini, M., de Cabo, R. & Ferrucci, L. *A rat epigenetic clock recapitulates phenotypic aging and co-localizes with heterochromatin.* Elife 9, (2020).](https://elifesciences.org/articles/59201)
- [Crowther, M. D., Dolton, G., Legut, M., Caillaud, M. E., Lloyd, A., Attaf, M., Galloway, S. A. E., Rius, C., **Farrell, C. P.**, Szomolay, B., Ager, A., Parker, A. L., Fuller, A., Donia, M., McCluskey, J., Rossjohn, J., Svane, I. M., Phillips, J. D. & Sewell, A. K. *Genome-wide CRISPR-Cas9 screening reveals ubiquitous T cell cancer targeting via the monomorphic MHC class I-related protein MR1.* Nat. Immunol. 21, 178–185 (2020).](https://www.nature.com/articles/s41590-019-0578-8)
- [Snir, S., **Farrell, C.** & Pellegrini, M. *Human epigenetic ageing is logarithmic with time across the entire lifespan.* Epigenetics 14, 912–926 (2019).](https://www.tandfonline.com/doi/full/10.1080/15592294.2019.1623634)
- [Phillips, J., **Farrell, C.**, Wang, Y., Singal, A. K., Anderson, K., Balwani, M., Bissell, M., Bonkovsky, H., Seay, T., Paw, B., Desnick, R. & Bloomer, J. *Strong correlation of ferrochelatase enzymatic activity with Mitoferrin-1 mRNA in lymphoblasts of patients with protoporphyria.* Mol. Genet. Metab. 128, 391–395 (2019).](https://pubmed.ncbi.nlm.nih.gov/30391163/)
- [Orozco, L. D., **Farrell, C.**, Hale, C., Rubbi, L., Rinaldi, A., Civelek, M., Pan, C., Lam, L., Montoya, D., Edillor, C., Seldin, M., Boehnke, M., Mohlke, K. L., Jacobsen, S., Kuusisto, J., Laakso, M., Lusis, A. J. & Pellegrini, M. *Epigenome-wide association in adipose tissue from the METSIM cohort.* Human Molecular Genetics 27, 1830–1846 (2018).](https://pubmed.ncbi.nlm.nih.gov/29566149/)
- [Yaish, H. M., **Farrell, C. P.**, Christensen, R. D., MacQueen, B. C., Jackson, L. K., Trochez-Enciso, J., Kaplan, J., Ward, D. M., Salah, W. K. & Phillips, J. D. Two novel mutations in TMPRSS6 associated with iron-refractory iron deficiency anemia in a mother and child. Blood Cells Mol. Dis. 65, 38–40 (2017).](https://www.ncbi.nlm.nih.gov/pmc/articles/PMC5656052/)
- [**Farrell, C. P.**, Overbey, J. R., Naik, H., Nance, D., McLaren, G. D., McLaren, C. E., Zhou, L., Desnick, R. J., Parker, C. J. & Phillips, J. D. *The D519G Polymorphism of Glyceronephosphate O-Acyltransferase Is a Risk Factor for Familial Porphyria Cutanea Tarda.* PLoS One 11, e0163322 (2016).](https://pubmed.ncbi.nlm.nih.gov/27661980/)
- **Farrell, C. P.**, Parker, C. J. & Phillips, J. D. *Exome sequencing for molecular characterization of non-HFE hereditary hemochromatosis.* Blood Cells Mol. Dis. 55, 101–103 (2015).
